File: java/Chord.java
import java.util.*;

public class Chord {

  private String chordName;
  public String tone;

  public String fullName(){
    return this.chordName + " " + this.tone;
  }

  public Chord(String chordName){
    this.chordName = chordName;
  }

  public Chord(String chordName, String tone){
    this.chordName = chordName;
    this.tone = tone;
  }

  public Note[] majorNotes(){
    // Start with a new array to hold the notes of the chord
    Note[] notes = new Note[3];

    // First, push the root note
    notes[0] = new Note(this.chordName);

    // Push the third degree of the chord
    notes[1] = notes[0].nextWholeStep().nextWholeStep();

    // Push the 5th degree of the chord
    notes[2] = notes[1].nextHalfStep().nextWholeStep();

    // Return the array
    return notes;
  }

  public Note[] minorNotes(){
    // Start with a new array to hold the notes of the chord
    Note[] notes = new Note[3];

    // First, push the root note
    notes[0] = new Note(this.chordName);

    // Push the third degree of the chord
    notes[1] = notes[0].nextWholeStep().nextHalfStep();

    // Push the 5th degree of the chord
    notes[2] = notes[1].nextWholeStep().nextWholeStep();

    // Return the array
    return notes;
  }

  public String[] listMajorNotes(){
    // Array to hold the note names
    String[] notes = new String[3];

    // Save the major notes array
    Note[] majorNotes = this.majorNotes();

    // Fill the notes array with the names of the major notes
    for (int i = 0; i < majorNotes.length; i++){

      // Push only the note name into the array
      notes[i] = majorNotes[i].noteName;
    };

    // Return the array with the note names
    return notes;
  }

  public String[] listMinorNotes(){
    // Array to hold the note names
    String[] notes = new String[3];

    // Save the major notes array
    Note[] minorNotes = this.minorNotes();

    // Fill the notes array with the names of the minor notes
    for (int i = 0; i < minorNotes.length; i++){

      // Push only the note name into the array
      notes[i] = minorNotes[i].noteName;
    };

    // Return the array with the note names
    return notes;
  }

  public static Chord[] diagnose(String[] noteArray){
    // Variable to keep track of the high score set by the winning chord
    int score = 0;
    Chord[] currentWinner = new Chord[1];

    // Array of all possible notes
    String[] allNotes = new String[]{"A","A#","B","C","C#","D","D#","E","F","F#","G","G#"};

    // Loop through every note
    for (int i = 0; i < allNotes.length; i++){

      // Create a new chord for the current note
      Chord chord = new Chord(allNotes[i]);

      // Keep track of the highest variation score within the chord
      int personalScore = 0;

      // Array to keep track of notes in each variations
      String[][] variationsArray = new String[][]{chord.listMajorNotes(),chord.listMinorNotes()};

      // Array to keep track of names of each variation
      String[] variationsNamesArray = new String[]{"Major","Minor"};

      // Our object to keep track of which variation has the highest score
      Object[] scoreKeeper = new Object[2];

      // Go through each variation of the chord
      for (int j = 0; j < variationsArray.length; j++){

        // Keep track of each variation's score
        int variationScore = 0;

        // Loop through the variation's notes
        for (int x = 0; x < variationsArray[j].length; x++){

          // If this variation has a common note with the parameter passed to the function, increment it's score
          for (int y = 0; y < allNotes.length; y++) {
            if (allNotes[y] == variationsArray[j][x]) {
              System.out.println(allNotes[y] + " matched with " + variationsArray[j][x]);
              variationScore++;
              break;
            };
          };
        };

        if (variationScore > personalScore){
          scoreKeeper = new Object[]{new String(variationsNamesArray[j]), new Integer(variationScore)};
          personalScore = variationScore;
        };
      };

      // Compare this chord's highest variation score to the high score
      if ((Integer)scoreKeeper[1] > score){
        // If we have a new high score, update the chords tone/variation
        chord.tone = (String)scoreKeeper[0];
        System.out.println("score: " + score + ", new current winner (first condition) " + chord.fullName() + " with personal score of: " + scoreKeeper[1]);

        // Set the new current winner
        currentWinner = new Chord[]{chord};

        score = (Integer)scoreKeeper[1];
      } else if ((Integer)scoreKeeper[1] == score){
         // If we have a tie, add him to the current winner array
         chord.tone = (String)scoreKeeper[0];
         System.out.println("score: " + score + ", new current winner (second condition) " + chord.fullName() + " with personal score of: " + scoreKeeper[1]);
         Chord[] holder = new Chord[currentWinner.length];
         System.arraycopy(holder, 0, currentWinner, 0, currentWinner.length);
         currentWinner = new Chord[currentWinner.length + 1];
         for (int z = 0; z < holder.length; z++){
          currentWinner[z] = holder[z];
         };
         currentWinner[currentWinner.length - 1] = chord;
      };
    };

    // Return the current winner
    return currentWinner;
  }

  public static void main(String[] args){

    // String[] aMinorNotes = new Chord("A","Minor").listMinorNotes();
    // String[] cMajorNotes = new Chord("C","Major").listMajorNotes();

    // Chord chord = new Chord(args[0],args[1]);
    // System.out.println(chord.tone);
    // if (chord.tone == "Major"){
    //   notes = chord.listMajorNotes();
    // } else if (chord.tone == "Minor") {
    //   notes = chord.listMinorNotes();
    // } else {
    //   notes = new String[]{"Unknown Tonality"};
    // };
    // System.out.println("A Minor Notes: ");
    // for (int x = 0; x < aMinorNotes.length; x++){
    //   System.out.print(aMinorNotes[x] + " ");
    // };

    // System.out.println("C Major Notes:");
    // for (int x = 0; x < cMajorNotes.length; x++){
    //   System.out.print(cMajorNotes[x] + " ");
    // };


    System.out.println(Chord.diagnose(new String[]{"A","C","E"})[0]);
  }
}

File: javascripts/chord.js
var Chord = function(attributes){
  this.chordName = attributes["chordName"];
  this.tone = attributes["tone"];
};

Chord.prototype = {

  fullName: function(){
    // Simple function to return name and tone, e.g A Minor
    return this.chordName + " " + this.tone;
  },

  majorNotes: function(){
    // Return an array of the notes that make up the major chord. Start with an empty array
    var notes = [];

    // Push the root, which is the name of the chord
    notes.push(new Note(this.chordName));

    // Push the 3rd degree of the scale
    notes.push(notes[0].nextWholeStep().nextWholeStep());

    // Push the 5th degree of the scale
    notes.push(notes[1].nextHalfStep().nextWholeStep());

    // Return the array
    return notes;
  },

  minorNotes: function(){
    // Return an array of the notes that make up the minor chord. Start with an empty array
    var notes = [];

    // Push the root/name of the chord
    notes.push(new Note(this.chordName));

    // Push the flat 3rd degree
    notes.push(notes[0].nextWholeStep().nextHalfStep());

    // Push the 5th degree
    notes.push(notes[1].nextWholeStep().nextWholeStep());

    // Return the array
    return notes;
  },

  listMajorNotes: function(){
    // Function to print the names of the major notes
    var notes = [];

    // Save the major notes array
    var majorNotes = this.majorNotes();

    // Go through every item in the major notes array
    for (var i = 0; i < majorNotes.length; i++){

      // Push the note name into the array
      notes.push(majorNotes[i].noteName);
    };

    // Return the array
    return notes;
  },

  listMinorNotes: function(){
    // Function to print the names of the minor notes
    var notes = [];

    // Save the minor notes array
    var minorNotes = this.minorNotes();

    // Go through each item of the minor notes array
    for (var i = 0; i < minorNotes.length; i++){

      // Push the note name into the array
      notes.push(minorNotes[i].noteName);
    };

    // Return the array
    return notes;
  }



};

// Object to keep track of all known interval values for the different chords
Chord.allChordIntervals = {
  "Major": "WWHWWWH",
  "Minor": "WHWWWWH"
};

// Object to keep track of the different degrees that make up different chords
Chord.degrees = {};

Chord.diagnose = function(notesArray){

  // Keep track of the winning interval
  var winningIntervalScore = 0;
  var winningInterval = "";
  var winningChord = "";
      // Keep track of the current winning root note
    var winningRootScore = 0;
    var winningRoot = "";

  // Loop through each saved interval value in the Chord.intervals store
  for (key in Chord.allChordIntervals){

    // We will save the key as the interval name, and the interval string as intervals
    var currentIntervalName = key;
    var intervals = Chord.allChordIntervals[key];

    // Go through every note from the Note.intervals object so we cover every possibility
    for (i in Note.allNotes){
      var root = Note.allNotes[i];

      // Keep Track of this root's score
      var rootScore = 0;

      // Save the current note
      var currentNote = root;

      for (var x = 0; x < intervals.length; x++){

        // Save the chord name so we know the tonality at the end of the function
        var chordName = root + ":" + currentIntervalName

        // If the current note is in the notesArray, add to the note score
        if (notesArray.indexOf(currentNote) > -1) {
          rootScore ++;
        };

        // If we have a new winner, store it
        if (rootScore > winningRootScore){
          winningRootScore = rootScore;

          // Only update the winning note if it is different than the current root
          if (winningRoot != chordName){ winningRoot = chordName; };

        // if there is a tie, add the current note to the winning note
        } else if (rootScore === winningRootScore && winningRoot !== root){

          // if winningNote is already an array, push current note into it
          if (Object.prototype.toString.call( winningRoot ) === "[object Array]" && winningRoot.indexOf(chordName) === -1){
            winningRoot.push(chordName);

          // if winning note is not an array, turn it into an array with itself and the current note as it's items
          } else if (Object.prototype.toString.call( winningRoot ) !== "[object Array]" && winningRoot !== chordName) {
            winningRoot = [winningRoot, chordName];
          };
        };

        // Save the current chracter of the intervals string
        var currentInterval = intervals.charAt(x);

        // If the current intervale is a W for whole step, set the current note to the next whole step
        if (currentInterval === "W"){
          currentNote = new Note(currentNote).nextWholeStep().noteName;

        // If the current interval is a H for half step, set the current note to the next half step
        } else if (currentInterval === "H"){
          currentNote = new Note(currentNote).nextHalfStep().noteName;

        // Otherwise the interval is bad, just stop
        } else {
          return false;
        };
      };
    };

    if (winningRootScore > winningIntervalScore){
      winningIntervalScore = winningRootScore;
      winningInterval = currentIntervalName;
    } else if (winningRootScore === winningIntervalScore){
      if (Object.prototype.toString.call( winningInterval ) === "[object Array]" && winningRoot.indexOf(currentIntervalName) === -1){
        winningRoot.push(currentIntervalName);
      } else if (Object.prototype.toString.call( winningInterval ) !== "[object Array]") {
        winningInterval = [winningInterval, currentIntervalName];
      };
    };
  };

  return winningRoot;
};

Chord.evaluateDegrees = function(notesArray){

  // Keep our result in an object
  var results = {};

  // Diagnose the notes array and save the chords
  var chords = Chord.diagnose(notesArray);

  // Go through each root, line up the degrees
  for ( i in chords ){

    // Split the chord string to get the root and the tone, then retrieve the intervals from the allChordIntervals object
    var chordSplit = chords[i].split(":")
    var root = chordSplit[0];
    var tone = chordSplit[1];
    var intervals = Chord.allChordIntervals[tone];

    // Store the degrees that align with each root
    var degrees = [];
    var currentNote = root;

    for (x in intervals){
      if (notesArray.indexOf(currentNote) !== -1){
        degrees.push(parseInt(x)+1);
      };
      if (intervals[x] === "W"){
        currentNote = new Note(currentNote).nextWholeStep().noteName;
      } else if (intervals[x] === "H"){
        currentNote = new Note(currentNote).nextHalfStep().noteName;
      } else {
        return false;
      };
    };
    results[root + " " + tone] = degrees;
  };
  return results;
};

